### ISSUE TITLE
Poisson Reconstruction for "mixed" data

### ISSUE THREAD

**Issue by artem-sham** [mkazhdan/PoissonRecon#245]
Dear Professor,

I'm working with 3D measurements taken with different 3D scanners, each of these scanners' clouds can be successfully processed separately (with appropriate settings), and such separate processing produces nice surfaces.
However, if all these different types of clouds merged together and processed with parameters providing maximal details (for the most detailed scanner type), then all of the parts exhibit some sort of artifacts. Mostly, the surfaces have "excessive details".
All of the measurements are quite dense, but at the same time they are noisy and noise model is available for each scanner.

I'd like to use this "uncertainty" information for each measurement alongside the position and normal direction in order to reconstruct a surface with just the right amount of details in each of the scanners' areas.

Could you please point me into a right direction of how to modify the code/algorithm/settings in order to achieve this.

Thank You!

**Comment by mkazhdan** [comment]
You could try doing the following:
In your point cloud, encode your confidence of the quality of samples in the magnitude of the normals. (Higher-quality points should have normals with magnitude close to one, lower-quality points should have normals with smaller magnitude.)
Then you can use the --confidence and --confidenceBias flags to get the reconstruction algorithm to give points with less confidence less "weight". This would have the effect of treating dense but noisy samples as lower-resolution.

**Comment by artem-sham** [comment]
Thank you for your kind reply!

I was thinking about changing the weights (encoded into lengths of the normal vectors) for the data obtained from each individual scanner to different values.
But if I, for example, make the weights to be 0.5 instead of default value of 1.0 for one scanner's data, then wouldn't it require just roughly twice more measurements from that scanner to be able to go as deep as the original points with the unity weights would go?
Do I understand the current implementation of weight correctly?
If the user decides to scan some particular area much more densely, then this area will inevitably be represented on a much finer level, regardless of the actual scanner's accuracy limitation (for example, if it's known that there's no need to go finer than 1.0 mm for one particular scanner)?
If my understanding is correct, do you think there is a way to limit such density-driven detailization/resolution selection to different levels for different scanners (under the assumption that I can't control how dense a user decides to make scans with each scanner)?

**Comment by mkazhdan** [comment]
There are couple of things at play here:
1. If you specify that you are using normal confidence, then the magnitude of the normal determines the fraction of the point that contributes. In particular, points are added up to a depth such that, roughly, there "samplesPerNode" samples within that node. So if only a fraction of the point contributes, it is harder to get deeper nodes. (So I think you would want to make the normal have length 0.25 if you wanted to force things to one depth coarser.
2. In general, you can force the finest resolution by using the --width parameter. (Though that may not be what you want if you want some scans to go deeper than others.)

**Comment by artem-sham** [comment]
Yes, I'd prefer every scan to go only as deep as required for this particular type of scanner.
I was considering adding nonlinear "deactivation" function for weight of the points added to the `DensityEstimator` in `_addWeightContribution` [function](https://github.com/mkazhdan/PoissonRecon/blob/master/Src/FEMTree.WeightedSamples.inl#L155).
For example I can check if the added point's level (known from scanner type) is not coarser than the depth of the node, to which this point is added, and if it is indeed too coarse for current node - then do not add anything to the density.
I hope this will change the decision about "selected" depth in the `_getSampleDepthAndWeight` function later on (when we compare the density to _samplesPerNode_ parameter), and will not allow to select the depth finer than required.
But I'm afraid that such modification will break everything, because I do not fully understand how the components of the pipeline affect each other.
Do you think it's worth trying to make such a change, and what other parts of the pipeline might be affected by this?

**Comment by mkazhdan** [comment]
One thing that could be done (though the released code-base doesn't quite support it now) is have multiple point streams, one for each scanner type, have each generate its own density estimator and right-hand-side constraints (up to a pre-determined stream-specific depth) and then merge all of those into a single right-hand-side constraint that is then solved.

The challenging thing with your proposal is that after lines 651/652 when the points are read in, the point stream gets deleted in line 656 and all you are left with is an amalgamated point-set, with the set of points falling into the same octree node represented by a single average sample. This would make it hard to tease apart which points came from where when constructing the density estimator or defining the vector field.

**Comment by artem-sham** [comment]
I didn't have time to try the separate points streams approach for each scanner.
It sounds more reasonable, but requires more work from my side in my application.
But I almost succeeded with the adaptation of `DensityEstimator` to account for each points's "uncertainty level" (scale).
I had to remove the step of creating "samples" from point stream, and to work with the point stream directly on each subsequent stage (density calculation, normals accumulation, adding point interpolation constraints and extracting isoValue).
The "coarse" data does not go deeper than required (which is what I was looking for).
But there are some problems with two "global" parameters - isoValue and screening factor (balancing term for point-based constraints relative to the normals-based constraints).
When I merge together different non-overlapping parts of one scene (scanned with different scanners) there is a noticeable difference with the surfaces that I get by processing each part of the scene separately.
It seems that the "normalized area" (computed as `_pointWeightSum / weightSum` during [splatting of normal vectors](https://github.com/mkazhdan/PoissonRecon/blob/master/Src/FEMTree.inl#L453)) gets too big for the whole scene (because the approximated area is much bigger than the one estimated from only one part of the scene), and eventually the point-based constraints get over-amplified. Reducing the screening weight does help, but this balance gets off again when I consider different mix of the scene parts, and I have to manually adjust it for each individual case (even though the main part which I use for comparison always remains the same, and all the other parts are actually quite far away from it).
Ideally, I'd like to avoid such influence from the distant parts of the scene and take into account only "local" information for determining the normal-vs-point balancing factor and isoValue.
What would be the right approach to make these two values more "local"? If it makes sense at all?

**Comment by artem-sham** [comment]
I'll close the ticket now, because I was able to finish my approach and even brought back the agglomeration of points into samples. The problem for me was that for one scanner I had big `weightSum` due to lots of points and small `_pointWeightSum` due to small area, for the other I had small `weightSum` due to sparse data and huge  `_pointWeightSum` due to large area. When combined together the resulting screening weight (= `_pointWeightSum / weightSum`) was too large for fine scanner and too small for coarse scanner, leading to over-amplification of fine details and over-smoothing of coarse details. I solved it by adaptively adjusting the screening weight.
![spf_old_](https://user-images.githubusercontent.com/3049290/211525120-d9c8d361-4a09-40c9-b40a-d7443e4a7adf.png)
![spf_new_](https://user-images.githubusercontent.com/3049290/211525231-6bf0d742-30c4-4bcc-852d-0d01cdfec43c.png)
